content/posts/cross-compile-query-json.mdx

```diff
@@ -0,0 +1,252 @@
+
+These expressions are called [Variants](https://reasonml.github.io/docs/en/variant), are one of the best features of Reason, some functional programming communities call it *ADT* (Algebraic Data Type) which is also correct.
+
+In this case those are *GADT* (Generic Algebraic Data Type) a kind of *ADTs* since they contain arguments and those can be recursive. The entire `query-json` program representation it's one big recursive Variant.
+
+Following with a more complex example, let's parse `.store.books | filter(.price > 10)`:
+
+```reason
+Pipe(
+    Pipe(
+        Key("store"),
+        Key("books")
+    ),
+    Filter(
+        Pipe(
+            Key("price"),
+            Literal(Number(10.))
+        )
+    )
+)
+```
+
+You can see more examples in the [parsing tests](https://github.com/davesnx/query-json/blob/master/test/ParseTest.re)
+
+### Compiling
+Compile recieve the *AST* expression and transforms it to code. It's a big recurive [pattern match](https://reasonml.github.io/docs/en/pattern-matching), which is another great feature of Reason and looks something like this:
+
+```reason
+let rec compile = (expression, json) => {
+  switch (expression) {
+  | Empty => empty
+  | Keys => keys(json)
+  | Key(key, opt) => member(key, opt, json)
+  | Index(idx) => index(idx, json)
+  | Head => head(json)
+  | Tail => tail(json)
+  | Length => length(json)
+  /* [...] */
+}
+```
+
+On the left site are defined all the possible `Variants` and on the right site the operations. Those operations transforms the JSON. Here is where `map`, `filter`, `reduce`, `index`, etc... are implemented.
+
+### Running
+The easier part, the compile step give us back a curried function that expects a json as only argument, we just apply the function to this JSON and print the result.
+
+This example only describe the happy path, in reality the parsing and compilation steps return a `Result` type which allows handling errors more composable.
+
+# Distribution
+Now we already covered how it works internally and a little overview of the architecture. Now let's dive on how developers can use it in their machines.
+
+## How to cross-compilation to binary
+query-json uses `esy` as a package manager that helps managing either native OCaml and Reason from opam registry or JavaScript packages from npm registry; and `dune` as a build system.
+
+All of the build process and their tests runs on our CI in Github Actions, which allows to run Mac, Windows and Linux images to compile with the OCaml compiler to different architectures.
+
+Once the build is success it pushes the binary into Github Releases and npm registry.
+
+## How compile to the Web
+The compilation to JavaScript is the sweet section of this blog post, since we are using OCaml (under the hood while writting Reason), it allows us to compile directlty to JavaScript using [js_of_ocaml](https://github.com/ocsigen/js_of_ocaml).
```

Comment:
```suggestion
The compilation to JavaScript is the sweet section of this blog post, since we are using OCaml (under the hood while writing Reason), it allows us to compile directly to JavaScript using [js_of_ocaml](https://github.com/ocsigen/js_of_ocaml).
```